modules/bm_sim/include/bm_sim/P4Objects.h

```diff
@@ -61,7 +61,9 @@ class P4Objects {
   explicit P4Objects(std::ostream &outstream = std::cout)
       : outstream(outstream) { }
 
-  int init_objects(std::istream *is, int device_id = 0, size_t cxt_id = 0,
+  int init_objects(std::istream *is,
+                   LookupStructureFactory * lookup_factory,
```

Comment:
That is a good point. But a bunch of code actually relies on the fact that there is only one `Switch` instance per binary. For example, action primitive registration, which lets you register target-specific primitives.
However, I do believe removing all static elements is a worthy goal. I think we can leave it as it is in your Pull Request. If more elements become customizable, we should probably bundle them in a `SwitchConfiguration` object or something (just made up that class), which would include your lookup factory. This way we would avoid ending up with huge signatures for methods like `init_objects`.